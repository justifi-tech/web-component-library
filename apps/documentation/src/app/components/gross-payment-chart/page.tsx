export default function GrossPaymentChartPage() {
  return (
    <div>
      <h1>Gross Payment Chart</h1>
      <p className="lead">
        A data visualization component for displaying payment analytics and trends.
      </p>

      <div className="mt-4">
        <h2>Overview</h2>
        <p>
          The Gross Payment Chart component provides interactive charts and graphs to visualize 
          payment data, revenue trends, and business performance metrics.
        </p>
      </div>

      <div className="mt-4">
        <h2>Usage</h2>
        <div className="bg-light p-3 rounded">
          <pre><code>{`import { JustifiGrossPaymentChart } from '@justifi/webcomponents';

function App() {
  return (
    <JustifiGrossPaymentChart 
      businessId="bus_123456789"
      chartType="line"
      dateRange={{ start: '2024-01-01', end: '2024-12-31' }}
    />
  );
}`}</code></pre>
        </div>
      </div>

      <div className="mt-4">
        <h2>Props</h2>
        <div className="table-responsive">
          <table className="table table-striped">
            <thead>
              <tr>
                <th>Prop</th>
                <th>Type</th>
                <th>Required</th>
                <th>Description</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td><code>businessId</code></td>
                <td>string</td>
                <td>Yes</td>
                <td>Business identifier for data filtering</td>
              </tr>
              <tr>
                <td><code>chartType</code></td>
                <td>string</td>
                <td>No</td>
                <td>Chart type: 'line', 'bar', 'area', 'pie'</td>
              </tr>
              <tr>
                <td><code>dateRange</code></td>
                <td>object</td>
                <td>No</td>
                <td>Date range for data display</td>
              </tr>
              <tr>
                <td><code>metrics</code></td>
                <td>array</td>
                <td>No</td>
                <td>Metrics to display in the chart</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>

      <div className="mt-4">
        <h2>Features</h2>
        <ul>
          <li>Multiple chart types (line, bar, area, pie)</li>
          <li>Interactive data visualization</li>
          <li>Real-time data updates</li>
          <li>Customizable date ranges</li>
          <li>Export charts as images</li>
          <li>Responsive design for all devices</li>
        </ul>
      </div>

      <div className="mt-4">
        <h2>Examples</h2>
        
        <h3>Bar Chart</h3>
        <div className="bg-light p-3 rounded">
          <pre><code>{`<JustifiGrossPaymentChart 
  businessId="bus_123456789"
  chartType="bar"
  metrics={['gross_payments', 'net_payments', 'fees']}
/>`}</code></pre>
        </div>

        <h3>Area Chart with Custom Range</h3>
        <div className="bg-light p-3 rounded">
          <pre><code>{`<JustifiGrossPaymentChart 
  businessId="bus_123456789"
  chartType="area"
  dateRange={{ 
    start: '2024-06-01', 
    end: '2024-06-30' 
  }}
  height={400}
/>`}</code></pre>
        </div>
      </div>
    </div>
  );
} 