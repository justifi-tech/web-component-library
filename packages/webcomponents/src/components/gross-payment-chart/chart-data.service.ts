import { Api } from '../../api';
import { config } from '../../../config';
import { GrossVolumeReponseType } from '../../api/GrossVolume';

export interface IChartDataService {
  fetchChartData(
    accountId: string,
    authToken: string,
  ): Promise<GrossVolumeReponseType>;
}

export class ChartDataService implements IChartDataService {
  async fetchChartData(
    accountId: string,
    authToken: string,
  ): Promise<GrossVolumeReponseType> {
    const api = Api(authToken, config.proxyApiOrigin);
    const endpoint = `account/${accountId}/reports/gross_volume`;
    return api.get(endpoint);
  }
}