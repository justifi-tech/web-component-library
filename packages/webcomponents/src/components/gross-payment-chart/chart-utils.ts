import { GrossVolumeReportDate } from "../../api/GrossVolume";
import { formatCurrency, formatDisplayDate } from "../../utils/utils";

export const generateChartOptions = (
  total: number,
  dates: GrossVolumeReportDate[],
  endDate: string
) => {
  return ({
    type: 'bar',
    options: {
      plugins: {
        legend: {
          display: false
        },
        title: {
          display: true,
          text: ['Trailing 30 Days', 'Gross Payments', formatCurrency(total)],
          position: 'top',
          align: 'start'
        },
        tooltip: {
          displayColors: false,
          intersect: false,
          callbacks: {
            label: (context) => {
              let index = context.dataIndex;
              let date = formatDisplayDate(dates[index].date, endDate);
              let value = formatCurrency(dates[index].value)
              return [date, value]
            },
          }
        }
      },
      scales: {
        x: {
          grid: {
            drawOnChartArea: false,
            drawTicks: false
          },
          ticks: {
            callback: (index) => {
              if (index === 0 || index === dates.length - 1) {
                return formatDisplayDate(dates[index].date, endDate);
              }
            }
          }
        },
        y: {
          display: false
        }
      }
    },
    data: {
      labels: dates.map(() => ''),
      datasets: [
        {
          label: 'Gross Volume by Date',
          data: dates.map((date) => date.value)
        }]
    },
  });
}