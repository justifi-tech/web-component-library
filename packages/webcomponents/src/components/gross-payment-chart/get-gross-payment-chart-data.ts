import { ComponentErrorSeverity } from '../../api/ComponentError';
import { getErrorCode, getErrorMessage } from '../../api/services/utils';

export const makeGetGrossPaymentChartData =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchGrossVolumeChartData(id, authToken);

      if (!response.error) {
        onSuccess(response.data);
      } else {
        const errorMessage = getErrorMessage(response.error);
        const code = getErrorCode(response.error?.code);
        onError({
          error: errorMessage,
          code,
          severity: ComponentErrorSeverity.ERROR,
        });
      }
    } catch (error) {
      const code = getErrorCode(error?.code);
      const errorMessage = getErrorMessage(error);
      onError({
        error: errorMessage,
        code,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  };
