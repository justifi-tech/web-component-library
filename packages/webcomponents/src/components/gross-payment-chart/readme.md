# justifi-gross-payment-chart



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description | Type     | Default     |
| ----------- | ------------ | ----------- | -------- | ----------- |
| `accountId` | `account-id` |             | `string` | `undefined` |
| `authToken` | `auth-token` |             | `string` | `undefined` |


## Events

| Event         | Description | Type                          |
| ------------- | ----------- | ----------------------------- |
| `error-event` |             | `CustomEvent<ComponentError>` |


## Dependencies

### Depends on

- [gross-payment-chart-core](.)

### Graph
```mermaid
graph TD;
  justifi-gross-payment-chart --> gross-payment-chart-core
  style justifi-gross-payment-chart fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
