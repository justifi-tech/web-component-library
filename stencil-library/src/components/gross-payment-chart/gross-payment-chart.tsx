import { Component, Host, Prop, State, Watch, h } from '@stencil/core';
import { Api, IApiResponseCollection } from '../../api';
import { Chart, BarController, Colors, BarElement, CategoryScale, LinearScale, Legend, Tooltip} from 'chart.js'
// import { mockGrossVolumeReport } from '../../api/mockData/mockGrossVolumeReport';
import { formatCurrency, formatDisplayDate } from '../../utils/utils';

interface GrossVolumeReportDate {
  date: string, value: number
}

interface GrossVolumeReport {
  total: number,
  dates: GrossVolumeReportDate[]
}

@Component({
  tag: 'justifi-gross-payment-chart',
  styleUrl: 'gross-payment-chart.scss',
  shadow: true,
})
export class GrossPaymentChart {
  chart: Chart
  chartRef: HTMLCanvasElement

  @Prop() accountId: string;
  @Prop() authToken: string;
  @State() data: GrossVolumeReport;
  @State() total: number;
  @State() dates: GrossVolumeReportDate[];
  @State() endDate: string;
  @State() loading: boolean;
  @State() errorMessage: string;

  @Watch('accountId')
  @Watch('authToken')
  updateOnPropChange() {
    this.fetchData();
  }

  connectedCallback() {
    this.fetchData();
  }
  
  componentDidRender() {
    this.renderChart();
  }

  async fetchData(): Promise<void> {
    if (!this.accountId || !this.authToken) {
      this.errorMessage = "Can not fetch any data without an AccountID and an AuthToken";
      this.loading = false;
      return;
    }

    this.loading = true;

    const api = Api(this.authToken, process.env.PRIVATE_API_ORIGIN);
    const endpoint = `account/${this.accountId}/reports/gross_volume`;
    const response: IApiResponseCollection<GrossVolumeReport> = await api.get(endpoint);

    if (!response.error) {
      console.log(response.data);
      this.total = response?.data.total;
      this.dates = response?.data.dates.reverse();
      this.endDate = this.dates[this.dates.length - 1].date;
      console.log(this.endDate);
    }
  }

  renderChart() {
    if (this.chart) {
      this.chart.update()
    } else if (this.chartRef && this.endDate) {
      Chart.register(
        Colors,
        BarController,
        BarElement,
        CategoryScale,
        LinearScale,
        Legend,
        Tooltip
      )
      this.chart = new Chart(
        this.chartRef.getContext("2d"),
        {
          type: 'bar',
          options: {
            plugins: {
              tooltip: {
                displayColors: false,
                callbacks: {
                  label: (context) => {
                    let index = context.dataIndex;
                    let date = formatDisplayDate(this.dates[index].date, this.endDate);
                    let value = formatCurrency(this.dates[index].value)
                    return [date, value]
                  },
                }
              }
            },
            scales: {
              x: {
                display: false
              },
              y: {
                display: false
              }
            }
          },
          data: {
            labels: this.dates.map(() => ''),
            datasets: [
              {
                label: 'Gross Volume Report',
                data: this.dates.map((date) => date.value),
                minBarLength: 5
                
              }]
          },
        }
      );
      this.chart.render()
    }
  }

  render() {
    return (
      <Host>
        <canvas id="chart" ref={(elem) => this.chartRef = elem} />
      </Host>
    );
  }

}
