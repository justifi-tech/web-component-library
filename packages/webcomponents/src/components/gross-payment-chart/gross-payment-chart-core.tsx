import { Component, Event, EventEmitter, Host, Prop, State, Watch, h } from '@stencil/core';
import { Chart, BarController, Colors, BarElement, CategoryScale, LinearScale, Legend, Tooltip, Title, ChartConfiguration } from 'chart.js'
import { GrossVolumeReport } from '../../api/GrossVolume';
import { generateChartOptions } from './chart-utils';
import { ErrorState, LoadingState } from '../details/utils';
import { ComponentError } from '../../api/ComponentError';

Chart.register(
  Colors,
  BarController,
  BarElement,
  CategoryScale,
  LinearScale,
  Legend,
  Tooltip,
  Title
)
@Component({
  tag: 'gross-payment-chart-core',
  styleUrl: 'gross-payment-chart.css',
})

export class GrossPaymentChartCore {
  chart: Chart
  chartRef: HTMLCanvasElement

  @Prop() getGrossPayment: Function;
  @State() grossVolumeReport: GrossVolumeReport;
  @State() loading: boolean = true;
  @State() errorMessage: string = '';
  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  componentDidLoad() {
    if (this.getGrossPayment) {
      this.fetchData();
    }
  }

  @Watch('getGrossPayment')
  propChanged() {
    if (this.getGrossPayment) {
      this.fetchData();
    }
  }

  @Watch('grossVolumeReport')
  grossVolumeReportChanged(grossVolumeReport: GrossVolumeReport) {
    if (grossVolumeReport) {
      if (!this.chart) {
        this.initChart(grossVolumeReport);
      } else {
        this.updateChartData(grossVolumeReport);
      }
    }
  }

  async fetchData() {
    this.loading = true;

    this.getGrossPayment({
      onSuccess: (data: GrossVolumeReport) => {
        this.loading = false;
        this.grossVolumeReport = data;
      },
      onError: ({ error, code, severity }) => {
        this.loading = false;
        this.errorMessage = error;

        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        })
      }
    });
  }

  initChart(grossVolumeReport: GrossVolumeReport) {
    const { dates, total } = grossVolumeReport;
    const reversedDates = dates.reverse();
    const endDate = reversedDates[dates.length - 1].date;
    const chartOptions = generateChartOptions(total, reversedDates, endDate) as ChartConfiguration;
    this.chart = new Chart(this.chartRef.getContext('2d'), chartOptions);
  }

  updateChartData(grossVolumeReport: GrossVolumeReport) {
    const newData = grossVolumeReport.dates.map((date) => date.value);
    this.chart.data.datasets.forEach((dataset) => {
      dataset.data = newData;
    });
    this.chart.update();
  }

  disconnectedCallback() {
    if (this.chart) {
      this.chart.destroy();
    }
  }

  render() {
    return (
      <Host>
        {this.loading && LoadingState()}
        {!this.errorMessage ? <canvas id="chart" ref={(elem) => this.chartRef = elem} /> : ErrorState(this.errorMessage)}
      </Host>
    );
  }
} 
