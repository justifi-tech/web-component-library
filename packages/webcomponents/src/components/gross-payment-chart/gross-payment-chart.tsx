import { Component, Event, EventEmitter, Prop, State, Watch, h } from '@stencil/core';
import { ReportsService } from '../../api/services/reports.service';
import { makeGetGrossPaymentChartData } from './get-gross-payment-chart-data';
import { ErrorState } from '../details/utils';
import { ComponentError, ComponentErrorCodes } from '../../api/ComponentError';

@Component({
  tag: 'justifi-gross-payment-chart',
  shadow: true,
})
export class GrossPaymentChart {
  @Prop() accountId: string;
  @Prop() authToken: string;

  @State() getGrossPayment: Function;
  @State() errorMessage: string = null;

  @Event() errorEvent: EventEmitter<ComponentError>;

  componentWillLoad() {
    this.initializeGetGrossPayment();
  }

  @Watch('accountId')
  @Watch('authToken')
  propChanged() {
    this.initializeGetGrossPayment();
  }

  private initializeGetGrossPayment() {
    if (this.accountId && this.authToken) {
      this.getGrossPayment = makeGetGrossPaymentChartData({
        id: this.accountId,
        authToken: this.authToken,
        service: new ReportsService(),
      });
    } else {
      this.errorMessage = 'Account ID and Auth Token are required';
      this.errorEvent.emit({
        errorCode: ComponentErrorCodes.MISSING_PROPS,
        message: this.errorMessage,
      });
    }
  }

  handleErrorEvent = (event: CustomEvent<ComponentError>) => {
    this.errorMessage = event.detail.message;
    this.errorEvent.emit(event.detail);
  };

  render() {
    if (this.errorMessage) {
      return ErrorState(this.errorMessage);
    }

    return (
      <gross-payment-chart-core getGrossPayment={this.getGrossPayment} onErrorEvent={this.handleErrorEvent} />
    );
  }
}


