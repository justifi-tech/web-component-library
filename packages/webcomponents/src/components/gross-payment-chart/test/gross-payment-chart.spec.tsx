import { h } from "@stencil/core";
import { newSpecPage } from "@stencil/core/testing";
import { GrossPaymentChart } from "../gross-payment-chart";
import { GrossPaymentChartCore } from "../gross-payment-chart-core";
import { ReportsService } from "../../../api/services/reports.service";
jest.mock("../../../api/services/reports.service");

describe('GrossPaymentChart', () => {
  it('renders an error when no accountId and authToken is passed', async () => {
    const page = await newSpecPage({
      components: [GrossPaymentChart, GrossPaymentChartCore],
      template: () => <justifi-gross-payment-chart />,
    });

    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('renders an error when no accountId is passed', async () => {
    const page = await newSpecPage({
      components: [GrossPaymentChart, GrossPaymentChartCore],
      template: () => <justifi-gross-payment-chart auth-token="abc" />,
    });

    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('renders an error when no authToken is passed', async () => {
    const page = await newSpecPage({
      components: [GrossPaymentChart, GrossPaymentChartCore],
      template: () => <justifi-gross-payment-chart account-id="abc" />,
    });

    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('emits an error event when no accountId and authToken is passed', async () => {
    const errorEvent = jest.fn();
    const page = await newSpecPage({
      components: [GrossPaymentChart, GrossPaymentChartCore],
      template: () => <justifi-gross-payment-chart onErrorEvent={errorEvent} />,
    });

    await page.waitForChanges();
    expect(errorEvent).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'missing-props',
          message: 'Account ID and Auth Token are required',
        }
      })
    );
  });

  it('emits an error event when fetch fails', async () => {
    ReportsService.prototype.fetchGrossVolumeChartData = jest.fn().mockReturnValue({
      "error": {
        "code": "not_authenticated",
        "message": "Not Authenticated"
      }
    });

    const errorEvent = jest.fn();

    const page = await newSpecPage({
      components: [GrossPaymentChart, GrossPaymentChartCore],
      template: () => (
        <justifi-gross-payment-chart
          account-id="abc"
          auth-token="abc"
          onErrorEvent={errorEvent}
        />
      ),
    });

    await page.waitForChanges();

    expect(errorEvent).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'fetch-error',
          message: 'Not Authenticated',
        }
      })
    );
  });

  it('emits an error event when error occurs', async () => {
    ReportsService.prototype.fetchGrossVolumeChartData = jest.fn()
      .mockRejectedValue(new Error('Network error'));

    const errorEvent = jest.fn();

    const page = await newSpecPage({
      components: [GrossPaymentChart, GrossPaymentChartCore],
      template: () => (
        <justifi-gross-payment-chart
          account-id="abc"
          auth-token="abc"
          onErrorEvent={errorEvent}
        />
      ),
    });

    await page.waitForChanges();

    expect(errorEvent).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'fetch-error',
          message: 'Network error',
        }
      })
    );
  });
});
