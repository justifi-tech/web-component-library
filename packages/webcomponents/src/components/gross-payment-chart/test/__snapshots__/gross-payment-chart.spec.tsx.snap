// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GrossPaymentChart renders an error when no accountId and authToken is passed 1`] = `
<justifi-gross-payment-chart>
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" style="font-size: 1.2rem;">
      <p part="text color font-family">
        Account ID and Auth Token are required
      </p>
    </main>
  </template>
</justifi-gross-payment-chart>
`;

exports[`GrossPaymentChart renders an error when no accountId is passed 1`] = `
<justifi-gross-payment-chart auth-token="abc">
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" style="font-size: 1.2rem;">
      <p part="text color font-family">
        Account ID and Auth Token are required
      </p>
    </main>
  </template>
</justifi-gross-payment-chart>
`;

exports[`GrossPaymentChart renders an error when no authToken is passed 1`] = `
<justifi-gross-payment-chart account-id="abc">
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" style="font-size: 1.2rem;">
      <p part="text color font-family">
        Account ID and Auth Token are required
      </p>
    </main>
  </template>
</justifi-gross-payment-chart>
`;
